### PR TITLE
fix(rebuild): don't run lifecycle scripts twice on linked deps

### LINKED ISSUE #2905: [BUG] "npm rebuild" runs any symlinked package's lifecycle scripts twice.
### Current Behavior:

A package dependency installed as a symlink will get its lifecycle scripts (such as `postinstall`) run twice when running the `npm rebuild` subcommand of `npm`.

### Expected Behavior:

Lifecycle scripts should probably only be run once.

### Steps To Reproduce:

Set up a minimal/dummy dependency with lifecycle scripts:
- `mkdir test && cd test && npm init -y`
- `mkdir dependency && cd dependency && npm init -y`
- Add this as a preinstall script for the `dependency` dummy package: `"preinstall": "echo 'You will see this twice!'`

Add the dummy package with lifecycle scripts as a symlinked dependency:
- `cd ..`
- `npm install ./dependency`

Run `npm rebuild` and verify that the lifecycle script runs twice:
- `npm rebuild --foreground-scripts`

Example output:

```console
% npm rebuild --foreground-scripts

> dependency@1.0.0 preinstall
> echo 'You will see this twice!'

You will see this twice!

> dependency@1.0.0 preinstall
> echo 'You will see this twice!'

You will see this twice!
rebuilt dependencies successfully
```

### Environment:

- OS: macOS Catalina 10.15.7
- Node: v15.12.0
- npm: 7.6.3

### PR DESCRIPTION
**Pull request by wraithgar**
Fixes https://github.com/npm/cli/issues/2905

### CODE REVIEW

**Comment by wraithgar** [comment]
While we're looking at lifecycle scripts + linked deps we may want to see if https://github.com/npm/cli/issues/3918 is related enough to address in the same PR

File: workspaces/arborist/lib/arborist/rebuild.js [hunk 1]
```diff
@@ -136,27 +136,26 @@ module.exports = cls => class Builder extends cls {
...
-    if (!this[_ignoreScripts]) {
-      await this[_runScripts]('preinstall')
-    }
-    if (this[_binLinks] && type !== 'links') {
-      await this[_linkAllBins]()
-    }
-
-    // links should also run prepare scripts and only link bins after that
-    if (type === 'links') {
+    // links should run prepare scripts and only link bins after that
```
Comment (ruyadorno) [inline workspaces/arborist/lib/arborist/rebuild.js:139]: the comment sounds off now 🤔 maybe just remove it

**Review by ruyadorno — COMMENTED** [review]
it would be nice to have a test that throws if a lifecycle script runs a second time, that can now pass with this change 😁

**Comment by wraithgar** [comment]
> it would be nice to have a test that throws if a lifecycle script runs a second time, that can now pass with this change 😁

I'm gonna need help w/ that one and planned on pairing w/ someone for this.

**Review by ruyadorno — APPROVED** [review]
Let's land this in time for today's release and backlog an issue to work on a test later on.

ref: https://github.com/npm/statusboard/issues/439

**Comment by n8agrin** [comment]
There is suspicion that this PR caused lifecycle scripts in workspaces to stop working in 8.5.4. Bug is https://github.com/npm/cli/issues/4552

**Pull request merged**